test/hotspot/jtreg/compiler/c2/irTests/TestVectorizeURShiftSubword.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package compiler.c2.irTests;
+
+import compiler.lib.ir_framework.*;
+import java.util.Random;
+import jdk.test.lib.Asserts;
+import jdk.test.lib.Utils;
+
+/*
+ * @test
+ * @bug 8283307
+ * @key randomness
+ * @summary Auto-vectorization enhancement for unsigned shift right on signed subword types
+ * @requires vm.cpu.features ~= ".*simd.*"
+ * @library /test/lib /
+ * @run driver compiler.c2.irTests.TestVectorizeURShiftSubword
+ */
+
+public class TestVectorizeURShiftSubword {
+
+    private static final Random RANDOM = Utils.getRandomInstance();
+
+    final private static int NUM = 300;
+
+    private short[] shorta = new short[NUM];
+    private short[] shortb = new short[NUM];
+    private byte[] bytea = new byte[NUM];
+    private byte[] byteb = new byte[NUM];
+
+    private static final int[] SPECIALS = {
+        0, 0x1, 0x8, 0xF, 0x3F, 0x7C, 0x7F, 0x80, 0x81, 0x8F, 0xF3, 0xF8, 0xFF,
+        0x38FF, 0x3FFF, 0x8F8F, 0x8FFF, 0x7FF3, 0x7FFF, 0xFF33, 0xFFF8, 0xFFFF, 0xFFFFFF,
+        Integer.MAX_VALUE, Integer.MIN_VALUE
+    };
+
+    public static void main(String[] args) {
+        TestFramework.run();
+    }
+
+    @Test
+    @IR(counts = {IRNode.LOAD_VECTOR, ">0", IRNode.RSHIFT_VB, ">0", IRNode.STORE_VECTOR, ">0"})
+    public void testByte0() {
+        for(int i = 0; i < NUM; i++) {
+            byteb[i] = (byte) (bytea[i] >>> 3);
+        }
+    }
+
+    @Test
+    @IR(counts = {IRNode.LOAD_VECTOR, ">0", IRNode.RSHIFT_VB, ">0", IRNode.STORE_VECTOR, ">0"})
+    public void testByte1() {
+        for(int i = 0; i < NUM; i++) {
+            byteb[i] = (byte) (bytea[i] >>> 24);
+        }
+    }
+
+    @Test
+    @IR(failOn = {IRNode.LOAD_VECTOR, IRNode.RSHIFT_VB, IRNode.STORE_VECTOR})
+    public void testByte2() {
+        for(int i = 0; i < NUM; i++) {
+            byteb[i] = (byte) (bytea[i] >>> 25);
+        }
+    }
+
+    @Test
+    @IR(counts = {IRNode.LOAD_VECTOR, ">0", IRNode.RSHIFT_VS, ">0", IRNode.STORE_VECTOR, ">0"})
+    public void testShort0() {
+        for(int i = 0; i < NUM; i++) {
+            shortb[i] = (short) (shorta[i] >>> 10);
+        }
+    }
+
+    @Test
+    @IR(counts = {IRNode.LOAD_VECTOR, ">0", IRNode.RSHIFT_VS, ">0", IRNode.STORE_VECTOR, ">0"})
+    public void testShort1() {
+        for(int i = 0; i < NUM; i++) {
+            shortb[i] = (short) (shorta[i] >>> 16);
+        }
+    }
+
+    @Test
+    @IR(failOn = {IRNode.LOAD_VECTOR, IRNode.RSHIFT_VS, IRNode.STORE_VECTOR})
+    public void testShort2() {
+        for(int i = 0; i < NUM; i++) {
+            shortb[i] = (short) (shorta[i] >>> 17);
+        }
+    }
+
+    @Test
+    public void checkTest() {
+        testByte0();
+        for (int i = 0; i < bytea.length; i++) {
+            Asserts.assertEquals(byteb[i], (byte) (bytea[i] >>> 3));
```

Comment:
This test won't work as expected if `(byte) (bytea[i] >>> 3)` is also vectorized.
So another question: can we make sure `(byte) (bytea[i] >>> 3)` wouldn't be vectorized during testing?

Reply:
Yes, the argument `(byte) (bytea[i] >>> 3)` of the function call won't be vectorized. Even if the function call is inlined, https://github.com/openjdk/jdk/blob/339005dbc99e94ed094612c7b34eb0c93ca1f8c1/test/lib/jdk/test/lib/Asserts.java#L200, any control flow statements can't be vectorized by SLP https://github.com/openjdk/jdk/blob/339005dbc99e94ed094612c7b34eb0c93ca1f8c1/src/hotspot/share/opto/superword.cpp#L125. Besides I suppose the function call is hard to be inlined.

Reply:
> Besides I suppose the function call is hard to be inlined.

Then I would suggest running with `-XX:-Inline`.